src/vm/ssa.nit

```diff
@@ -0,0 +1,1140 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Copyright 2015 Julien Pagès <julien.pages@lirmm.fr>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Compute Single-Static Assignment form from an AST
+module ssa
+
+import variables_numbering
+import astbuilder
+
+redef class VirtualMachine
+
+	# The currently analyzed APropdef
+	var current_propdef: APropdef
+
+	redef fun new_frame(node, mpropdef, args)
+	do
+		# Save the current propdef
+		if node isa APropdef then current_propdef = node
+
+		return super
+	end
+end
+
+# Represent a sequence of the program
+# A basic block is composed of several instruction without a jump
+class BasicBlock
+	# First instruction of the basic block
+	var first: ANode is noinit
+
+	# Last instruction of the basic block
+	var last: ANode is noinit
+
+	# Direct successors
+	var successors = new Array[BasicBlock]
+
+	# Direct predecessors
+	var predecessors = new Array[BasicBlock]
+
+	# Parts of AST that contain a read to a variable
+	var read_sites = new Array[AVarFormExpr]
+
+	# Parts of AST that contain a write to a variable
+	var write_sites = new Array[AVarFormExpr]
+
+	# Parts of AST that contain a variable access (read or write)
+	var variables_sites = new Array[AExpr]
+
+	# The iterated dominance frontier of this block
+	# i.e. the set of blocks this block dominate directly or indirectly
+	var dominance_frontier: Array[BasicBlock] = new Array[BasicBlock] is lazy
+
+	# Self is the old block to link to the new,
+	# The two blocks are not linked if the current ends with a `AReturnExpr` or `ABreakExpr`
+	# i.e. self is the predecessor of `successor`
+	# `successor` The successor block
+	fun link(successor: BasicBlock)
+	do
+		# Do not link the two blocks if the current block end with a return
+		if last isa AReturnExpr then return
+
+		if last isa ABreakExpr then return
+
+		successors.add(successor)
+		successor.predecessors.add(self)
+	end
+
+	# Self is the old block to link to the new
+	# i.e. self is the predecessor of `successor`
+	# `successor` The successor block
+	fun link_special(successor: BasicBlock)
```

Comment:
This one will not return if we try to link two blocks and if the first is ending with a break.

This is used to link a block ending with a break to the block where the break "jumps".